modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/SnapshotRestoreProcess.java

```diff
@@ -0,0 +1,799 @@
+
+        Throwable err = opCtx0.err.get();
+
+        if (err != null)
+            return new GridFinishedFuture<>(err);
+
+        if (!allNodesInBaselineAndAlive(opCtx0.nodes))
+            return new GridFinishedFuture<>(new IgniteCheckedException(OP_REJECT_MSG + "Server node(s) has left the cluster."));
+
+        GridFutureAdapter<Boolean> retFut = new GridFutureAdapter<>();
+
+        try {
+            Collection<StoredCacheData> ccfgs = opCtx0.cfgs.values();
+
+            // Ensure that shared cache groups has no conflicts before start caches.
+            for (StoredCacheData cfg : ccfgs) {
+                if (!F.isEmpty(cfg.config().getGroupName()))
+                    ensureCacheAbsent(cfg.config().getName());
+            }
+
+            if (log.isInfoEnabled()) {
+                log.info("Starting restored caches " +
+                    "[requestID=" + opCtx0.reqId + ", snapshot=" + opCtx0.snpName +
+                    ", caches=" + F.viewReadOnly(ccfgs, c -> c.config().getName()) + ']');
+            }
+
+            ctx.cache().dynamicStartCachesByStoredConf(ccfgs, true, true, false, null, true, opCtx0.nodes).listen(
+                f -> {
+                    if (f.error() != null) {
+                        log.error("Unable to start restored caches [requestID=" + opCtx0.reqId +
+                            ", snapshot=" + opCtx0.snpName + ']', f.error());
+
+                        retFut.onDone(f.error());
+                    }
+                    else
+                        retFut.onDone(true);
+                }
+            );
+        } catch (IgniteCheckedException e) {
+            log.error("Unable to restore cache group(s) from snapshot " +
+                "[requestID=" + opCtx0.reqId + ", snapshot=" + opCtx0.snpName + ']', e);
+
+            return new GridFinishedFuture<>(e);
+        }
+
+        return retFut;
+    }
+
+    /**
+     * @param reqId Request ID.
+     * @param res Results.
+     * @param errs Errors.
+     */
+    private void finishCacheStart(UUID reqId, Map<UUID, Boolean> res, Map<UUID, Exception> errs) {
+        SnapshotRestoreContext opCtx0 = opCtx;
+
+        if (opCtx0 == null || !reqId.equals(opCtx0.reqId))
+            return;
+
+        Exception failure = checkFailure(errs, opCtx0, res.keySet());
+
+        if (failure == null) {
+            finishProcess();
+
+            return;
+        }
+
+        if (U.isLocalNodeCoordinator(ctx.discovery()))
+            rollbackRestoreProc.start(reqId, new SnapshotRestoreRollbackRequest(reqId, failure));
+    }
+
+    /**
+     * Check the response for probable failures.
+     *
+     * @param errs Errors.
+     * @param opCtx Snapshot restore operation context.
+     * @param respNodes Set of responding topology nodes.
+     * @return Error, if any.
+     */
+    private Exception checkFailure(Map<UUID, Exception> errs, SnapshotRestoreContext opCtx, Set<UUID> respNodes) {
+        Exception err = F.first(errs.values());
```

Comment:
If you return only the first exception to the user (the `restore` future will be completed with this exception, right?) probably you should log other exceptions which you are skipping.

Reply:
It seems we should log local failures locally, I think it's better to add additional logging on each exception catching, what do you think?